### PR TITLE
Alternative to #12395.

### PR DESCRIPTION
**Pull request by sbordet**
Alternative to #12395
The main difference is assuming that the `Runnable` passed to `Content.Source.demand(Runnable)` is non-blocking unless explicitly declared so. This leaves most of the application code using lambdas and method references unchanged.

Deprecated usages of `CompletableFuture` APIs in `Content.Source`.

### CODE REVIEW

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java [hunk 1]
```diff
@@ -542,6 +547,12 @@ default long getLength()
          * {@code Runnable} are serialized and a callback for {@code demand} call is
          * not invoked until any previous {@code demand} callback has returned.
          * Thus the {@code Runnable} should not block waiting for a callback of a future demand call.</p>
+         * <p>The {@code Runnable} callback SHOULD be implemented with non-blocking code.
+         * In case it needs to be implemented with blocking code, it MUST be signaled to the implementation
+         * by using the following code:</p>
```
Comment (gregw) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:552]:
```suggestion
         * by using the following code (or similar using an Invocable):</p>
```

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java [hunk 2]
```diff
@@ -980,7 +978,10 @@ public void demand(Runnable demandCallback)
                 {
                     if (httpChannelState._onContentAvailable != null)
                         throw new IllegalArgumentException("demand pending");
-                    httpChannelState._onContentAvailable = demandCallback;
+                    if (demandCallback instanceof Invocable)
+                        httpChannelState._onContentAvailable = demandCallback;
+                    else
+                        httpChannelState._onContentAvailable = new ReadyTask(InvocationType.NON_BLOCKING, demandCallback);
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:984]: This is the core of this alternative proposal....  but I don't think it has achieved much as there are still 22 other files that are modified as you attempt to pass an Invocable rather than a plain Runnable.  So this PR is making pretty much the same changes as the original anyway.  

Changing the contract here doesn't win very much, nor does it help with the other 43 implementations of demand that all need to be checked and/or modified.

**Review by gregw — CHANGES_REQUESTED** [review]
I'm not convinced by this approach.

Whilst I'm not opposed to switching the assumption for demand callback from BLOCKING to NON_BLOCKING,  I think we should do so only if it gives us a big advantage.   In this case I don't think it is giving us much.

Specifically, even with the changed assumption, we still feel the need to change many classes to ensure the passed `Runnable` is an `Invocable`, which if we do, means we do not need to change the contract.

Also, there are 44 implementations of `demand`, so there is not a single place that we can change this assumption and it will be hard to enforce.   So we are going to have to change both the demand implementation and the callers of demand, which doubles the work needed.

I think the least surprise is for us to keep the Invocable contract as is (assume blocking unless stated otherwise) and to fix the objects that we pass in (as both this PR and the original do).

I agree that CF's are still not the best, and I'm happy to move away from them were we can, but that is kind of orthogonal to both PRs.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java (hunk 2 above)
Comment (sbordet) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:984]: It is not about the change we make to Jetty files.

It is about the changes that applications need to do.
For example, with the other PR, CometD would need to be changed. With this PR, no.

Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:984]: Can you point to the changes needed in CometD?   Are they really needed?  Are there demand callbacks in cometd that are tagged as blocking, but ultimately do nothing but wake up another blocked thread?   If not, then they are not a problem.

The invocation type of our callbacks has always key, but it is only vital if it is used in the conversion from async to blocking APIs, which is typically done by our utility classes.    The fact that we messed up a few utility classes and forgot to pass on the invocation type should not mean that we change the contract.  If we assume non-blocking and call blocking code, we will starve the thread pool just the same.

Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:984]: @sbordet the more I think about this, the more I'm convinced this is not an issue with the `demand` API/contract.    There is nothing wrong with passing blocking tasks to `demand` and there is nothing wrong with assuming blocking unless told otherwise.  The mistake we made in a few recent utility classes was to block a Thread and to wait for a wakeup from a blocking task that was likely to come from the same thread pool as the original blocked thread.    This is just a basic programming error and the fact that the `demand` API was used is not really relevant, as the same issue would occur if `Executor.execute` was used.

Our previous/existing utility classes (`Blocker`, `SharedBlockingCallback`, `BlockingCallback`, `HttpInput` etc.) were all written correctly with the knowledge that if you block a thread, then you need to be very careful about how the event that wakes it up is scheduled and that it cannot be dependent on the wakeup of the blocked thread.

The fact that we made some programming errors in our refactor of `HttpInput` and the new `CompletableFuture` based utilities is on us, and not on the `demand` API.   We made some error, which are easy to fix (done in the original PR).     Sure others can make similar errors: async programming is hard.  But changing the contract of `demand` so our errors are glossed over, is not a fix.  It actually makes it worse, as we may now dispatch the last available thread to a callback that is validly blocking, but that we decided to declare non-blocking without any indication to the caller.... just on the assumption that if they happened to make the same programming mistake that we did, then we'd work around it.  If they have not made the same mistake, then this PR will lock the server.

**Comment by sbordet** [comment]
@gregw I think we're on similar conclusions, that some of the utility classes using `CompletableFuture` exposed the problem.

As for the demand `Runnable` to be blocking by default, I think that is fine for these reasons:

* It aligns with how we treat `Handler.handle()` -- blocking.
To be explicit, a request may arrive with all content already available, we call `handle()` which reads asynchronously, never hits a `null` chunk, and upon the last chunk it performs a blocking JDBC call; similarly, a request may arrive, have no content yet, so `handle()` would read a `null` chunk, and register a demand `Runnable` which when invoked would read all the chunks and upon the last perform the JDBC call.
So `Handler.handle()` and the demand `Runnable.run()` should be equivalent with respect to their `InvocationType`.
* Defaulting to non-blocking has the problem that `SerializedInvoker.Link`'s invocation type could not be hard-coded to blocking like it is now.
This PR changes it to look up the `Runnable` invocation type, but it's not right if a subsequent task is blocking.

On the other hand, perhaps most of the demand `Runnable` implementations are non-blocking, so it would be nice to avoid the extra ceremony to specify that, so assuming non-blocking would be a better default. 
However, I have no real-world evidence of how many demand `Runnable` implementations are non-blocking.

In summary, I think there is unrelated goodness in this PR (e.g. fixing some of the missing `executeImmediately()`), both PRs must pay attention to the fact that the demand `Runnable` may specify the `InvocationType`, and propagate it if the demand `Runnable` is wrapped.

As for use of `CompletableFuture`, I am for deprecating its usages where not appropriate, but I would avoid introducing custom `CompletableFuture` to try to overcome the problem: it was a mistake, we deprecate it.
Had we used `CompletionStage` instead of `CompletableFuture`, we would not have had the problem.

**Comment by gregw** [comment]
@sbordet we are in agreement with almost everything.... except when you say:

> I would avoid introducing custom `CompletableFuture` to try to overcome the problem: it was a mistake, we deprecate it.

I think we have to fix CF usage in the next releases. At best, we can only deprecate existing usage, so the API will remain for sometime.    It is entirely fixable (if somewhat fragile to do so), so we should fix it.   If we can deprecate and replace it as well, then great, but we have to fix it.

Currently my extended CF is only used by `ContentSourceCompletableFuture`,  which is then used by `FormFields`, `MultiPartFormData` and `MultiPartByteRanges`.  So a single moderately simple fix is addressing 3 of our most key use-cases.    We could just do the fix in `ContentSourceCompletableFuture` rather than in an explicit `InvocableCompletableFuture`, but I think that exacerbates the fragility of the fix.   The issue is, that with a fat interface on CF then the fixed methods will be "lost" if they are done in another class.  By making an explicit CF class, we contain the work around we need into one place... even if deprecated and scheduled for ultimate removal.     I don't particularly care if the fix adapts dynamically from the `andThen` methods or if they throw ISE or IAE (see my last comment on the other issue).


OK, I also don't understand when you say:
> Had we used CompletionStage instead of CompletableFuture, we would not have had the problem

Our problem is when we have a callback that is doing the adaption from async to blocking and is actually a non-blocking wakeup for an already blocked thread.   That is the `CF.get()` API which we use extensively to map our non-blocking parsers of several content-type to the blocking servlet APIs.  So we need the `CF.get()` methods, which are not provided by `CompletionStage`.  Adding our own get API doesn't help as we will still have the duality of usage of these utility classes: sometimes used asynchronously with `andThen` and sometimes used blocking with `get()` - and we need to adjust the invocation type accordingly.  Remember, that there is nothing wrong with a blocking demand callback, so long as it is not just a wakeup of a blocked thread.

So how would we fix this is we didn't have the CF interface?    We have fully async implementations of forms, and multiparts that can be used either async or blocking.    If we gave them only an async `Promise` based API, then we'd still need to check the `InvocationType` of that promise, and utility classes that converted the Promise into a blocking API would still need to ensure they were non-blocking.     Thus the "fix" would look exactly the same as what the other PR has done for CF, except it would not be on a fat API and thus less fragile.   I guess we could also use `Invocable.Task` in the new API, to make it clear we are checking the invocation type, although that is a bit of a pain for the valid use-cases of passing a real blocking callback.     In fact, thinking about it, I'm not 100% convinced that deprecating the CF is worth it (see below).

So, to move forward, can you:
- merge the goodness from this PR into the other one, which can be focused on fixing the APIs that we have.
- review the fix on ContentSourceCompletableFuture (using the extended CF) in the other PR.  Comment on which of the variants is your preference (or none).
- open another draft PR in which we can experiment with alternatives to the CF based APIs.   Only once we have found something better should we deprecate the CF APIs in that PR.

**Comment by gregw** [comment]
Actually, now I'm not entirely convinced with the CF fix either.  See discussion coming soon in the other PR....

**Comment by sbordet** [comment]
Closing, as the good changes from this PR have been ported to #12395 in 43232f588f154ab205ae0d74c820a1975de73e8d, and there is agreement that defaulting to blocking behavior is better.
